eth/backend.go

```diff
@@ -695,21 +734,47 @@ func (s *Ethereum) handleWhitelistCheckpoint(ctx context.Context, first bool) er
 		return ErrBorConsensusWithoutHeimdall
 	}
 
-	// Create a new checkpoint verifier
-	verifier := newCheckpointVerifier(nil)
-	blockNums, blockHashes, err := ethHandler.fetchWhitelistCheckpoints(ctx, bor, verifier, first)
+	// Create a new bor verifier, which will be used to verify checkpoints and milestones
+	verifier := newBorVerifier(nil)
+	blockNum, blockHash, err := ethHandler.fetchWhitelistCheckpoint(ctx, bor, s, verifier)
 	// If the array is empty, we're bound to receive an error. Non-nill error and non-empty array
 	// means that array has partial elements and it failed for some block. We'll add those partial
 	// elements anyway.
-	if len(blockNums) == 0 {
+	if err != nil {
 		return err
 	}
 
-	// Update the checkpoint whitelist map.
-	for i := 0; i < len(blockNums); i++ {
-		ethHandler.downloader.ProcessCheckpoint(blockNums[i], blockHashes[i])
+	ethHandler.downloader.ProcessCheckpoint(blockNum, blockHash)
+
+	return nil
+}
+
+// handleMilestone handles the milestone mechanism.
+func (s *Ethereum) handleMilestone(ctx context.Context) error {
+	ethHandler := (*ethHandler)(s.handler)
+
+	bor, ok := ethHandler.chain.Engine().(*bor.Bor)
+	if !ok {
+		return ErrNotBorConsensus
+	}
+
+	if bor.HeimdallClient == nil {
```

Comment:
I believe we have something like `IsHeimdall` or something like that.

Reply:
No, We don't have IsHeimdall function linked to Bor struct.